Sources/MongoSwift/BSON/BSONDecoder.swift

```diff
@@ -116,7 +116,7 @@ internal class _BSONDecoder: Decoder {
 
     // Returns the data stored in this decoder as represented in a container keyed by the given key type.
     public func container<Key>(keyedBy type: Key.Type) throws -> KeyedDecodingContainer<Key> {
-        guard self.storage.topContainer != nil  else {
+        guard !(self.storage.topContainer is NSNull)  else {
```

Comment:
So, I know what you mean, but I kept this in just because apparently, from the code, the intended behavior for a `null` is to throw this error with a message tailored for `null`. If we remove this and merge it into the codepath for other `BSONValue` types, then we lose this distinction.

This isn't inherently a bad thing, but we should make sure that this change in behavior is what we want.

Reply:
Is this related to the outstanding questions in https://github.com/mongodb/mongo-swift-driver/pull/145#discussion_r235508617?

Reply:
Nope, I don't think so. `valueNotFound` is specifically for `nil`'s that correspond to actual `null` values. The questions in the comment are about a situation where we are not actually sure which kind of `nil` we are dealing with, but in this case, we know.

Reply:
Do I understand correctly that JSONDecoder would report `DecodingError.valueNotFound` only for a JSON `null` value, and `DecodingError._typeMismatch` for anything else that isn't a JSON `object` type? Is this related to JSONDecoder using `nil` for JSON `null`, or do they also use NSNull?

If this behavior isn't actually dictated, I personally don't see the benefit of treating BSON `null` differently in this case. Relying on the `DecodingError._typeMismatch` error handling below seems fine and would let us delete a bit of code here.

Looking at the [DecodingError documentation](https://developer.apple.com/documentation/swift/decodingerror), `valueNotFound` says:

> An indication that a nonoptional value of the given type was expected, but a null value was found.

Conversation in [SR-5156](https://bugs.swift.org/browse/SR-5156) leads me to believe this is actually thrown for `nil` values.

Reply:
So I think SR-5156 seems to have been resolved, and so looking at the current code in the Swift stdlib, it seems like they are doing what we are doing here by checking against `NSNull` and throwing essentially the same exact error as we are here. So it seems like their idea of finding a "null value" is `NSNull`?

I think you're right that if it is not really dictated/documented anywhere, we shouldn't need to spend the extra LoC on this particular case. I doubt it is a big issue either way, but perhaps it may end up being helpful to have some ability to distinguish between a null vs. a type mismatch amongst non-null types?

I think I will defer this to @kmahar again as our resident `Encoder`/`Decoder` expert to get her opinion on the matter.

Reply:
let's just use typeMismatch

Reply:
Done.